### PR TITLE
[MED-2670] Garbage collecting old revisions

### PR DESCRIPTION
**Pull request by micahnoland**
Hello @dokipen, @ddbenson, @dnelson, @silverlyra, 

Please review the following commits I made in branch 'micahnoland/revision_gc':

- **Garbage collecting old revisions** (0d1765869730afa5129c422086865787581f932f)

Related to [MED-2670](https://medium.atlassian.net/browse/MED-2670).

R=@dokipen
R=@ddbenson
R=@dnelson
R=@silverlyra

### CODE REVIEW

File: pkg/controller/releasemanager/releasemanager_controller.go [hunk 1]
```diff
@@ -606,3 +609,36 @@ func (r *ResourceSyncer) SyncVirtualService() error {
...
+	}
+
+	lastCurrent := 0
+	for i, incarnation := range incarnations {
+		releaseStatus := r.Instance.ReleaseStatus(incarnation)
+		if releaseStatus.CurrentPercent > 0 {
+			lastCurrent = i
+		} else {
+			expiration := releaseStatus.GitTimestamp.Add(time.Duration(incarnation.Spec.Release.TTL) * time.Second)
+			if lastCurrent < i && i > lastCurrent+incarnation.Spec.Release.GcBuffer && time.Now().After(expiration) {
```
Comment (dokipen) [inline pkg/controller/releasemanager/releasemanager_controller.go]: lsn't `lastCurrent < i` superfluous?

File: pkg/controller/revision/revision_controller.go [hunk 2]
```diff
@@ -281,6 +281,17 @@ func (r *ReconcileRevision) SyncReleaseManagersForRevision(revision *picchuv1alp
 				rm.Spec.Target = targetName
 				return nil
 			})
+
+			for _, rl := range rm.Status.Releases {
```
Comment (dokipen) [inline pkg/controller/revision/revision_controller.go]: don't we want to make sure *all* releaseManagers have expired the revision before deleting? Like increment a counter for every expired and compare counter to rm count? If equal, delete?

File: pkg/controller/revision/revision_controller.go [hunk 3]
```diff
@@ -281,6 +281,17 @@ func (r *ReconcileRevision) SyncReleaseManagersForRevision(revision *picchuv1alp
...
+				if rl.Tag == revision.Spec.App.Tag && rl.Expired == true {
+					err := r.DeleteRevision(revision)
```
Comment (dokipen) [inline pkg/controller/revision/revision_controller.go]: if err := fn(); err != nil {
}

File: pkg/controller/revision/revision_controller.go [hunk 4]
```diff
@@ -281,6 +281,17 @@ func (r *ReconcileRevision) SyncReleaseManagersForRevision(revision *picchuv1alp
...
+					if err != nil {
+						log.Error(err, "Failed to delete Revision")
+						return err
+					}
+				}
+			}
+
 			if err != nil {
```
Comment (dokipen) [inline pkg/controller/revision/revision_controller.go]: this error check should probably go above the for `rm.Status.Releases` block

File: pkg/controller/revision/revision_controller.go (hunk 2 above)
Comment (micahnoland) [reply pkg/controller/revision/revision_controller.go]: Right duh, I think this is fixed.

File: pkg/controller/releasemanager/releasemanager_controller.go (hunk 1 above)
Comment (micahnoland) [reply pkg/controller/releasemanager/releasemanager_controller.go]: Right, this was leftover from an edge case in my first attempt. Not needed.

**Pull request merged**